AWSS3Tests/AWSS3TestHelper.m

```diff
@@ -0,0 +1,141 @@
+//
+// Copyright 2010-2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License").
+// You may not use this file except in compliance with the License.
+// A copy of the License is located at
+//
+// http://aws.amazon.com/apache2.0
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+//
+
+#import "AWSS3TestHelper.h"
+#define AWS_TEST_BAH_INSTEAD true
+
+@interface AWSS3CreateBucketConfiguration()
+
++ (NSValueTransformer *)locationConstraintJSONTransformer;
+
+@end
+
+@interface AWSEndpoint()
+
++ (NSString *)regionNameFromType:(AWSRegionType)regionType;
+
+@end
+
+@implementation AWSS3TestHelper
+
++ (BOOL)createBucketWithName:(NSString *)bucketName
+                   andRegion:(AWSRegionType)regionType {
+    
+    AWSS3 *s3 = [AWSS3 defaultS3];
+    
+    AWSS3CreateBucketRequest *createBucketReq = [AWSS3CreateBucketRequest new];
+    createBucketReq.bucket = bucketName;
+    
+    if (regionType != AWSRegionUSEast1) {
+        NSString *regionName = [AWSEndpoint regionNameFromType:regionType];
+        NSValueTransformer *locationConstraintTransformer = [AWSS3CreateBucketConfiguration locationConstraintJSONTransformer];
+        NSNumber *constraint = [locationConstraintTransformer transformedValue:regionName];
+        AWSS3BucketLocationConstraint dupcon = constraint.integerValue;
+        AWSS3CreateBucketConfiguration *createBucketConfiguration = [AWSS3CreateBucketConfiguration new];
+        createBucketConfiguration.locationConstraint = dupcon;
+        createBucketReq.createBucketConfiguration = createBucketConfiguration;
+    }
+
+    
+    __block BOOL success = NO;
+    [[[s3 createBucket:createBucketReq] continueWithBlock:^id(AWSTask *task) {
+        if (task.error) {
+            success = NO;
+        } else {
+            success = YES;
+        }
+        return nil;
+    }] waitUntilFinished];
+    
+    [[NSRunLoop currentRunLoop] runUntilDate:[NSDate dateWithTimeIntervalSinceNow:10]];
+    
+    return success;
+}
+
++ (BOOL)deleteBucketWithName:(NSString *)bucketName {
+    AWSS3 *s3 = [AWSS3 defaultS3];
+    
+    AWSS3ListObjectVersionsRequest *listObjectVersionsRequest = [AWSS3ListObjectVersionsRequest new];
```

Comment:
Why can't we just invoke `deleteAllObjects` before deleting the bucket? What is special about the individual deletions in this method?

Reply:
Just deleting the objects will not delete the versions. And without deleting the version delete bucket will fail. For versioned buckets, when you attempt to delete objects, Amazon S3 inserts delete markers for all objects, but doesn't delete the object versions.
           
So we need to individually delete the version if present.